Enable custom return status codes

nickelization opened the pull request:
This PR allows callbacks to specify custom status codes to be returned as a result of a Clique command being run.

These changes also allow existing callbacks to remain the same - triggering them will cause clique:run to just return 'ok' as before. Callbacks can now also specify custom exit status by returning {exit_status, Status, Result} instead of just Result, which will cause clique:run to return {error, ExitStatus} instead. Finally, internal Clique errors will (for now) default to returning an error code of 1.

This is technically a slight API change, insofar as that it is now possible for clique:run to return {error, StatusCode} as well as 'ok'. However, none of our existing code has a problem with this, so this would only potentially affect third-party code outside of Basho. For anyone who doesn't care about return codes, it's a very simple change to work around, and for someone who _does_ care about return codes, this should be a welcome enhancement.

nickelization committed:
Fix clique:run type spec

This can currently only ever return 'ok'.

nickelization committed:
Fix clique_command:run type spec

Not sure how Dialyzer was even passing with the old spec...

nickelization committed:
Add support for different exit statuses

This change allows existing callbacks to remain the same - triggering
them will cause clique:run to just return 'ok' as before. Callbacks can
now also specify custom exit status by returning {Result, ExitStatus}
instead of just Result. Finally, errors will (for now) default to an
error code of 1.

nickelization committed:
Fix clash with callbacks returning {error, E}

Just matching on a 2-tuple wasn't specific enough - callbacks are
allowed to return {error, _} which is supposed to cause a corresponding
message to be formatted and printed by clique.

We could make it more specific by adding an is_integer guard, but we
probably should have been explicitly matching on something specific like
this in the first place.

nickelization committed:
Fix clique_command:run type spec

If the callback returns {error, _}, we will still pass it through back
to the caller.

Well, technically we'll pass anything back through, but error tuples are
intended to be valid and allowable in this case, whereas other things
might not be.

nickelization committed:
Default to exit status 1 for callback errors

If the callback returns an {error, _} tuple, but doesn't specify any
particular exit code, it makes more sense to default to 1 than to 0.

macintux commented:
@nickelization Could you please edit the description rather than appending the edit in the middle? It's a bit confusing, so while I think I understand the full impact of the change + change to the change, it'd be useful to see it laid out as if you had made no changes along the way.

macintux commented on src/clique.erl:
```diff
@@ -81,15 +81,23 @@ register_usage(Cmd, Usage) ->
 %% @doc Take a list of status types and generate console output
--spec print(err() | clique_status:status() | {clique_status:status(), string()}, [string()]) -> ok.
+-spec print(err() | clique_status:status() | {clique_status:status(), integer(), string()},
+            [string()]) -> ok.
```
`print` can return an error tuple, so this spec is incomplete.

macintux commented:
I'd really like to see a test involving a callback that returns an exit status before giving this a thumbs up.

nickelization commented:
@macintux Thanks for the review comments! Replies below:
- I edited the description to remove/reintegrate the "EDIT" comments into the main description. Thought it might be clearer to explicitly show the edit to the description given that the original commits are still present, and I added some changes after originally submitting the PR, but I can definitely see how it was a bit annoying/awkward to read it the way it was.
- Nice catch on the type spec. I fixed something else in the arguments, but didn't touch the return value at all. I'll add your fix too.
- Good call on adding a unit test. Clique really needs a lot more tests in general, and I'm worried I might end up having to add a bunch of additional infrastructure to write a test for this one specific thing...but I suppose that's work that has to get done eventually anyway. And, hopefully my gut feeling is wrong and it'll be easy :). I'll take a look into it this afternoon.

nickelization committed:
Fix and improve clique:print/1 type spec

nickelization committed:
Add basic_cmd_test

This is technically unrelated to the other work in this branch, but I
wanted to get a basic test running first before I try to test the
specifics of how error statuses get handled and returned.

nickelization committed:
Add unit test for clique exit status functionality

nickelization commented:
Okay @macintux, review changes are in :). Anything else?

nickelization committed:
Fix typo in comment

nickelization commented:
@borshop merge
